File: go/0014_longest_common_prefix.go
package main

import "bytes"

func longestCommonPrefix(strs []string) string {
	strslen := len(strs)
	if strslen == 0 {
		return ""
	}

	var b bytes.Buffer
	var ch string
	for i := 0; ; i++ {
		ch = ""
		for j := 0; j < strslen; j++ {
			if i >= len(strs[j]) {
				return b.String()
			}
			if ch == "" {
				ch = string(strs[j][i])
			} else if string(strs[j][i]) != ch {
				return b.String()
			}
		}
		b.WriteString(ch)
	}
}

File: python/0014_longest_common_prefix.py
class Solution:
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """

        if not strs:
            return ''

        end, shouldBreak = 0, False
        while True:
            for i, s in enumerate(strs):
                if end > len(s) - 1:
                    if len(s) == 0:
                        return ''

                    shouldBreak = True
                    break

                if i == 0:
                    c = s[end]
                elif s[end] != c:
                    shouldBreak = True
                    break
            if shouldBreak:
                break
            end += 1
        return strs[0][:end]
